aeneas/src/ir/VariantSolver.v3

```diff
@@ -151,79 +360,65 @@ class VariantNormalizer(nc: NormalizerConfig, rn: ReachabilityNormalizer, verbos
 	}
 	private def unboxUsingTaglessVariantNorm(rc: RaClass) {
 		var ofs = rc.orig.fields;
+
+		var normFields = getNormFieldsForCase(rc);
+		var shouldPack = rc.orig.packed;
+		var solver = VariantSolver.new(this, shouldPack, false);
+		var solvable = solver.solve([normFields]);
+
+		var types = Arrays.map(solver.curRep.extract(), getTypeFromScalarSet(_, shouldPack));
+		var vn = VariantNorm.new(rc.oldType, Tuple.newType(Lists.fromArray(types)), types, rc.variantFields, null);
+
+		vn.fieldRanges = rc.fieldRangesO;
+		vn.vecO = rc.vecO;
+		rc.variantNorm = vn;
+		setVariantFields(rc, 0, solver, shouldPack);
+
+		if (verbose) Terminal.put1("variant norm %q\n", rc.variantNorm.render);
+	}
+	private def getNormFieldsForCase(rc: RaClass) -> Array<Type> {
+		var ofs = rc.orig.fields;
 		var fields = Array<VariantField>.new(ofs.length);
-		var fieldRanges = Array<(int, int)>.new(ofs.length);
+		var fieldTypes = Vector<Type>.new();
 
+		rc.fieldRangesO = Array<(int, int)>.new(ofs.length);
+		rc.fieldRangesT = Array<(int, int)>.new(ofs.length);
 		var vecO = Vector<Type>.new();
-		var vecT = Vector<Type>.new();
 
 		for (i < ofs.length) {
-			var normStart = vecT.length, origStart = vecO.length;
 			var rf = rc.fields[i];
-			var indexes: Array<int>;
+			var startT = fieldTypes.length;
 
 			if (rf != null && rf.normIndex >= 0) {
 				var tn = fieldNorm(rf);
-				tn.addTo(vecT);
-
-				indexes = Array<int>.new(tn.size);
-				for (j < indexes.length) indexes[j] = normStart + j;
-				fields[i] = VariantField.new(tn, indexes);
+				tn.addTo(fieldTypes);
+				fields[i] = VariantField.new(tn, Array<int>.new(tn.size));
 			} else fields[i] = EMPTY_FIELD;
-			
+
+			var origStart = vecO.length;
 			var fieldType = ofs[i].fieldType.substitute(V3.getTypeArgs(rc.oldType));
```

Comment:
If you have the `RaField` you shouldn't need to do a type substitution. If you're doing a type substitution then it's usually a sign you should be working with `Ra..` entities instead. E.g. the type you end up with might not have been encountered (and processed) yet.

Reply:
I'll take a look at this and see if I can safely remove this. I think this was carried over from the original implementation of normalization using `DataNorm`s, but I don't know if it's doing anything at this point.

Reply:
We need this type substitution here in computing `vecO`, since `RaField`'s aren't generated for fields never encountered, but `vecO` is to help us normalize allocations of the class, and the arguments for unused fields aren't eliminated.